server/v2/stf/core_router_service.go

```diff
@@ -0,0 +1,124 @@
+package stf
+
+import (
+	"context"
+	"errors"
+	"fmt"
+	"strings"
+
+	"google.golang.org/protobuf/runtime/protoiface"
+
+	appmodulev2 "cosmossdk.io/core/appmodule/v2"
+	"cosmossdk.io/core/router"
+)
+
+// NewRouterService creates a router.Service which allows to invoke messages and queries using the msg router.
+func NewRouterService(queryRouterBuilder *MsgRouterBuilder, msgRouterBuilder *MsgRouterBuilder) router.Service {
+	queryRouter, err := queryRouterBuilder.Build()
+	if err != nil {
+		panic("cannot create queryRouter")
+	}
+
+	msgRouter, err := msgRouterBuilder.Build()
+	if err != nil {
+		panic("cannot create msgRouter")
+	}
+
+	return &routerService{
+		queryRouterService: &queryRouterService{
+			builder: queryRouterBuilder,
+			handler: queryRouter,
+		},
+		msgRouterService: &msgRouterService{
+			builder: msgRouterBuilder,
+			handler: msgRouter,
+		},
+	}
+}
+
+var _ router.Service = (*routerService)(nil)
+
+type routerService struct {
+	queryRouterService router.Router
+	msgRouterService   router.Router
+}
+
+// MessageRouterService implements router.Service.
+func (r *routerService) MessageRouterService() router.Router {
+	return r.msgRouterService
+}
+
+// QueryRouterService implements router.Service.
+func (r *routerService) QueryRouterService() router.Router {
+	return r.queryRouterService
+}
+
+var _ router.Router = (*msgRouterService)(nil)
+
+type msgRouterService struct {
+	builder *MsgRouterBuilder
+	handler appmodulev2.Handler
+}
+
+// CanInvoke returns an error if the given message cannot be invoked.
+func (m *msgRouterService) CanInvoke(ctx context.Context, typeURL string) error {
+	if typeURL == "" {
+		return errors.New("missing type url")
+	}
+
+	typeURL = strings.TrimPrefix(typeURL, "/")
+	if exists := m.builder.HandlerExists(typeURL); exists {
+		return fmt.Errorf("unknown request: %s", typeURL)
+	}
+
+	return nil
+}
+
+// InvokeTyped execute a message and fill-in a response.
+// The response must be known and passed as a parameter.
+// Use InvokeUntyped if the response type is not known.
+func (m *msgRouterService) InvokeTyped(ctx context.Context, msg, resp protoiface.MessageV1) error {
```

Comment:
```suggestion
func (m *msgRouterService) InvokeTyped(ctx context.Context, msg, resp *protoiface.MessageV1) error {
```
isn't this suppose to be a pointer?

Reply:
I think it's OK, interface types implicitly are pointers; runtime/server v1 usage: https://github.com/cosmos/cosmos-sdk/blob/3e633092203d90bf27c8be5fade08ea404ca358c/runtime/router.go#L73-L84

Reply:
I didn't give it much thought because we have the same usage in runtime/v1 router, but I don't think it will work as intended even as a pointer. That usage only works there due to the proto.Merge usages in the hybrid router I believe.

Reply:
see: https://github.com/cosmos/cosmos-sdk/pull/20349